{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "46b9b595",
   "metadata": {},
   "source": [
    "Perform NicheNet analysis\n",
    "================\n",
    "Maria Solovey\n",
    "2022-- DATE\n",
    "\n",
    "In this notebook, we perform a basic NicheNet analysis on a Seurat object. We start by creating a Seurat object from the normalized count matrix and cell annotation file. You can find the notebooks to preprocessing, filtering and UMAP as well as the details of the input dataset in XXX. \n",
    "\n",
    "\n",
    "The used NicheNet custom networks, ligand-target matrix and expression\n",
    "data of interacting cells can be downloaded from Zenodo. The NicheNet\n",
    "networks and ligand-target matrix at [![DOI](https://zenodo.org/badge/DOI/10.5281/zenodo.3260758.svg)](https://doi.org/10.5281/zenodo.3260758)\n",
    "and the raw single-cell data along with normalized and filtered at [![DOI](https://zenodo.org/badge/DOI/10.5281/zenodo.3531889.svg)](https://doi.org/10.5281/zenodo.3531889)\n",
    "\n",
    "\n",
    "This pipeline is adapted from the offical tutorial of NicheNet's on their GitHub page. \n",
    "\n",
    "[Perform NicheNet analysis starting from a Seurat object: step-by-step analysis\n",
    "](https://github.com/saeyslab/nichenetr/blob/master/vignettes/seurat_steps.md)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "87801534",
   "metadata": {},
   "source": [
    "# Prepare NicheNet analysis\n",
    "\n",
    "## Load required packages, read in the normalized count matrix of interacting cells and custom build of ligand-target prior model for NicheNet, ligand-receptor network and weighted integrated networks.\n",
    "\n",
    "To identify potential ligand-receptor interactions between two cell populations, the NicheNet approach requires both the NicheNet ligand-receptor network and weighted networks. The ligand-target matrix provides information on the potential of certain ligands to regulate the expression of specific target genes, enabling prioritization of possible interactions based on observed gene expression effects through NicheNet's ligand activity analysis. This prioritization helps to infer the affected target genes of the selected ligands."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "b4f9f658",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Attaching SeuratObject\n",
      "\n",
      "Warning message:\n",
      "“package ‘tidyverse’ was built under R version 4.2.1”\n",
      "── \u001b[1mAttaching packages\u001b[22m ────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────── tidyverse 1.3.2 ──\n",
      "\u001b[32m✔\u001b[39m \u001b[34mtibble \u001b[39m 3.1.8      \u001b[32m✔\u001b[39m \u001b[34mdplyr  \u001b[39m 1.0.10\n",
      "\u001b[32m✔\u001b[39m \u001b[34mtidyr  \u001b[39m 1.2.1      \u001b[32m✔\u001b[39m \u001b[34mstringr\u001b[39m 1.5.0 \n",
      "\u001b[32m✔\u001b[39m \u001b[34mreadr  \u001b[39m 2.1.3      \u001b[32m✔\u001b[39m \u001b[34mforcats\u001b[39m 0.5.2 \n",
      "\u001b[32m✔\u001b[39m \u001b[34mpurrr  \u001b[39m 1.0.0      \n",
      "Warning message:\n",
      "“package ‘forcats’ was built under R version 4.2.1”\n",
      "── \u001b[1mConflicts\u001b[22m ───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────── tidyverse_conflicts() ──\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mfilter()\u001b[39m masks \u001b[34mstats\u001b[39m::filter()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mlag()\u001b[39m    masks \u001b[34mstats\u001b[39m::lag()\n",
      "\n",
      "Attaching package: ‘igraph’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:dplyr’:\n",
      "\n",
      "    as_data_frame, groups, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:purrr’:\n",
      "\n",
      "    compose, simplify\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:tidyr’:\n",
      "\n",
      "    crossing\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:tibble’:\n",
      "\n",
      "    as_data_frame\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    decompose, spectrum\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:base’:\n",
      "\n",
      "    union\n",
      "\n",
      "\n",
      "Loading required package: circlize\n",
      "\n",
      "========================================\n",
      "circlize version 0.4.15\n",
      "CRAN page: https://cran.r-project.org/package=circlize\n",
      "Github page: https://github.com/jokergoo/circlize\n",
      "Documentation: https://jokergoo.github.io/circlize_book/book/\n",
      "\n",
      "If you use it in published research, please cite:\n",
      "Gu, Z. circlize implements and enhances circular visualization\n",
      "  in R. Bioinformatics 2014.\n",
      "\n",
      "This message can be suppressed by:\n",
      "  suppressPackageStartupMessages(library(circlize))\n",
      "========================================\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘circlize’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:igraph’:\n",
      "\n",
      "    degree\n",
      "\n",
      "\n",
      "Warning message:\n",
      "“package ‘R.utils’ was built under R version 4.2.2”\n",
      "Loading required package: R.oo\n",
      "\n",
      "Warning message:\n",
      "“package ‘R.oo’ was built under R version 4.2.1”\n",
      "Loading required package: R.methodsS3\n",
      "\n",
      "Warning message:\n",
      "“package ‘R.methodsS3’ was built under R version 4.2.1”\n",
      "R.methodsS3 v1.8.2 (2022-06-13 22:00:14 UTC) successfully loaded. See ?R.methodsS3 for help.\n",
      "\n",
      "R.oo v1.25.0 (2022-06-12 02:20:02 UTC) successfully loaded. See ?R.oo for help.\n",
      "\n",
      "\n",
      "Attaching package: ‘R.oo’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:R.methodsS3’:\n",
      "\n",
      "    throw\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:igraph’:\n",
      "\n",
      "    hierarchy\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:methods’:\n",
      "\n",
      "    getClasses, getMethods\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    attach, detach, load, save\n",
      "\n",
      "\n",
      "R.utils v2.12.2 (2022-11-11 22:00:03 UTC) successfully loaded. See ?R.utils for help.\n",
      "\n",
      "\n",
      "Attaching package: ‘R.utils’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:tidyr’:\n",
      "\n",
      "    extract\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:utils’:\n",
      "\n",
      "    timestamp\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    cat, commandArgs, getOption, isOpen, nullfile, parse, warnings\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘data.table’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:dplyr’:\n",
      "\n",
      "    between, first, last\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:purrr’:\n",
      "\n",
      "    transpose\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# libraries\n",
    "library(nichenetr)\n",
    "library(ggplot2)\n",
    "library(Seurat)\n",
    "library(tidyverse)\n",
    "library(igraph)\n",
    "require(circlize)\n",
    "library(R.utils)\n",
    "library(data.table) #to read gz file"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dda535fc",
   "metadata": {},
   "source": [
    "### Read in the expression data of interacting cells:\n",
    "The dataset used here is publicly available single-cell data from XXX. The data was processed, and filtered by applying XXX. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "fe2139c6",
   "metadata": {},
   "outputs": [],
   "source": [
    "input_dir <- \"../../../../../../results/data_preprocessing/Lasry/preprocessed/\"\n",
    "db_dir <- \"../../../../../../results/method_comparison/build_customDB/NicheNet/\"\n",
    "output_dir <- \"../../../../../../results/method_comparison/compare_results/NicheNet/\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "06953a65",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # load counts\n",
    "# print(\"load counts\")\n",
    "# counts <- read.table(gzfile(paste0(path_in,\"/counts_corr.csv.gz\")\n",
    "#                             )\n",
    "#                      ,sep = \",\"\n",
    "#                      ,row.names = 1\n",
    "#                      ,header = TRUE\n",
    "#                      )\n",
    "# # load counts\n",
    "\n",
    "counts <- fread(paste0(input_dir,\"counts_corr.csv.gz\"), header = TRUE,check.names=FALSE)\n",
    "counts <- as.data.frame(counts)\n",
    "rownames(counts) <- counts$gene_symbol\n",
    "counts <- counts[,-1]\n",
    "# head(str(counts))\n",
    "# print(str(counts))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "c9e31eb8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"load cell annotation\"\n"
     ]
    }
   ],
   "source": [
    "# load cell annotation\n",
    "print(\"load cell annotation\")\n",
    "anno_cells <- read.table(paste0(input_dir,\"anno_cells_corr.txt\")\n",
    "                         ,sep = \"\\t\"\n",
    "                         ,row.names = 1\n",
    "                         ,header = TRUE\n",
    "                         )\n",
    "# print(str(anno_cells))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "074c35b2",
   "metadata": {},
   "outputs": [],
   "source": [
    "srt=CreateSeuratObject(counts=counts, meta.data=anno_cells)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "53c467f9",
   "metadata": {},
   "source": [
    "#### Read in ligand-target prior model, ligand-receptor network and weighted integrated networks:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "d8987bb1",
   "metadata": {},
   "outputs": [],
   "source": [
    "ligand_target_matrix = readRDS(paste0(db_dir,\"custom_ligand_target_matrix.rds\"))\n",
    "lr_network = readRDS(paste0(db_dir,\"custom_lr_network.rds\"))\n",
    "weighted_networks = readRDS(paste0(db_dir,\"custom_weighted_networks.rds\"))\n",
    "weighted_networks_lr = weighted_networks$lr_sig %>% inner_join(lr_network %>% distinct(from,to), by = c(\"from\",\"to\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "e1258392",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       ".\n",
       "    AML healthy \n",
       "  21311   25391 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "srt@meta.data$health_status %>% table()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "7d5be848",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       ".\n",
       "    B    DC   Ery  Gran  HSPC  Mono    NK     T \n",
       " 4765  1634  1674  2332  3169 18004  3078 12046 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "srt@meta.data$cell_type %>% table()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "e239d63b",
   "metadata": {},
   "outputs": [],
   "source": [
    "Idents(srt) <- \"cell_type\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "92a8acc7",
   "metadata": {},
   "outputs": [],
   "source": [
    "cell_type_list <- unique(srt@meta.data$cell_type)\n",
    "sample_list <- unique(srt@meta.data$sample_ID)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "498b45a3",
   "metadata": {},
   "source": [
    "### 1. Define a “sender/niche” cell population and a “receiver/target”\n",
    "In this case study, we consider each cell type as reciever as well as a sender in a for loop. We consider a gene to be expressed when it is expressed in at least 10% of cells in one cluster. We have 8 cell types, and each of them is tested as a sender as well as reciever. This allows us to check the cell types of sender and reciever later on during the interpretation. \n",
    "\n",
    "### 2. Define a gene set of interest\n",
    "Here, the gene set of interest are the genes differentially expressed in each cell type after --AML diagnosis--. Thus, `AML` is the condition of interest, whereas the reference condition is annotated as `healthy`. The method to calculate the differential expression is here the standard Seurat Wilcoxon test. \n",
    "\n",
    "### 3. Define a set of potential ligands\n",
    "These are ligands that are expressed by the sender cell population and bind a receptor expressed by the reciever population.\n",
    "\n",
    "### 4. Perform NicheNet ligand activity analysis\n",
    "\n",
    "Once NicheNet outputs which ligands from one or more cell population(s) (“sender/niche”) are most likely to affect target gene expression in an interacting cell population (“receiver/target”), we extract the pairs from the `weighted_networks_lr`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "7e3c60a1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create an empty list to store the expressed genes for each sample and cell type\n",
    "expressed_genes_list <- list()\n",
    "\n",
    "# Iterate through each sample in the sample_list\n",
    "for (sample in sample_list){\n",
    "    \n",
    "    # Subset the Seurat object to the current sample\n",
    "    sample_subset <- subset(x = srt, subset = sample_ID == sample)\n",
    "    \n",
    "    # Get a list of unique cell types in the current sample\n",
    "    cell_types <- unique(sample_subset@meta.data$cell_type)\n",
    "    \n",
    "    # Iterate through each cell type in the current sample\n",
    "    for (cell_type in cell_types){\n",
    "        \n",
    "        # Call the function 'get_expressed_genes' to get the genes that are expressed in the current cell type\n",
    "        expressed_genes <- get_expressed_genes(cell_type, sample_subset, 0.10)\n",
    "        \n",
    "        # Convert the expressed genes to a vector and remove duplicates\n",
    "        expressed_genes <- expressed_genes %>% unlist() %>% unique()\n",
    "        \n",
    "        # Add the list of expressed genes for the current sample and cell type to the 'expressed_genes_list'\n",
    "        expressed_genes_list[[sample]][[cell_type]] <- expressed_genes\n",
    "        \n",
    "    }\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "93ef36c4",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "7618041e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create an empty list to store the differential expression tables for each cell type\n",
    "DE_tables <- list()\n",
    "\n",
    "# Iterate through each cell type in the cell_type_list\n",
    "for (cell_type in cell_type_list){\n",
    "    \n",
    "    # Subset the Seurat object to the current cell type\n",
    "    seurat_subset <- subset(srt, idents = cell_type)\n",
    "    \n",
    "    # Set the identity of the cells in the subset to the 'health_status' column\n",
    "    seurat_subset <- SetIdent(seurat_subset, value = seurat_subset[[\"health_status\"]])\n",
    "    \n",
    "    # Set the conditions of interest and reference for the differential expression analysis\n",
    "    condition_oi <- \"AML\"\n",
    "    condition_reference <- \"healthy\"\n",
    "    \n",
    "    # Call the function 'FindMarkers' to perform differential expression analysis and get the differential expression table\n",
    "    DE_table <- FindMarkers(object = seurat_subset, ident.1 = condition_oi, \n",
    "                                    ident.2 = condition_reference, min.pct = 0.10) %>% rownames_to_column(\"gene\")\n",
    "    \n",
    "    # Add the differential expression table for the current cell type to the 'DE_tables' list\n",
    "    DE_tables[[cell_type]] <- DE_table\n",
    "}\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "14262a89",
   "metadata": {},
   "source": [
    "This function is designed to transform a data frame into a new data frame in wide format, with the same structure as the output from a community tool. The reason for this is to ensure consistency in the structure of the results table, which can facilitate further analysis or comparison with other data sets.\n",
    "\n",
    "In particular, the input data frame is expected to have columns named \"interaction\", \"sample\", and \"weight\", representing interactions between entities in different samples. The function pivots the data frame from long format to wide format, with \"interaction\" as the ID column, \"sample\" as the column names, and \"weight\" as the values.\n",
    "\n",
    "The resulting data frame has the same number of rows as the input data frame, but the number of columns is equal to the number of unique values in the \"sample\" column. This structure is consistent with the output of a community tool, which typically provides a table of interactions with each sample represented as a separate column."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "27a3b316",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define a function 'transform_interaction_df' that takes a data frame as input\n",
    "transform_interaction_df <- function(df) {\n",
    "  \n",
    "  # Pivot the data frame to wide format, with 'interaction' as the ID column, \n",
    "  # 'sample' as the column names, and 'weight' as the values\n",
    "  df_new <- df %>%\n",
    "    pivot_wider(\n",
    "      id_cols = interaction,\n",
    "      names_from = sample,\n",
    "      values_from = c(\"weight\")\n",
    "    ) %>%\n",
    "    \n",
    "    # Reorder the columns to have 'interaction' as the first column\n",
    "    select(interaction, everything()) \n",
    "  \n",
    "  # Return the pivoted data frame\n",
    "  return(df_new)\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a47c70f6",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "50e842b6",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in evaluate_target_prediction(setting, ligand_target_matrix, ligands_position):\n",
      "“all target gene probability score predictions have same value”\n",
      "Warning message in cor(prediction, response):\n",
      "“the standard deviation is zero”\n",
      "Warning message in cor(prediction, response, method = \"s\"):\n",
      "“the standard deviation is zero”\n",
      "Warning message in evaluate_target_prediction(setting, ligand_target_matrix, ligands_position):\n",
      "“all target gene probability score predictions have same value”\n",
      "Warning message in cor(prediction, response):\n",
      "“the standard deviation is zero”\n",
      "Warning message in cor(prediction, response, method = \"s\"):\n",
      "“the standard deviation is zero”\n",
      "Warning message in evaluate_target_prediction(setting, ligand_target_matrix, ligands_position):\n",
      "“all target gene probability score predictions have same value”\n",
      "Warning message in cor(prediction, response):\n",
      "“the standard deviation is zero”\n",
      "Warning message in cor(prediction, response, method = \"s\"):\n",
      "“the standard deviation is zero”\n",
      "Warning message in evaluate_target_prediction(setting, ligand_target_matrix, ligands_position):\n",
      "“all target gene probability score predictions have same value”\n",
      "Warning message in cor(prediction, response):\n",
      "“the standard deviation is zero”\n",
      "Warning message in cor(prediction, response, method = \"s\"):\n",
      "“the standard deviation is zero”\n",
      "Warning message in evaluate_target_prediction(setting, ligand_target_matrix, ligands_position):\n",
      "“all target gene probability score predictions have same value”\n",
      "Warning message in cor(prediction, response):\n",
      "“the standard deviation is zero”\n",
      "Warning message in cor(prediction, response, method = \"s\"):\n",
      "“the standard deviation is zero”\n",
      "Warning message in evaluate_target_prediction(setting, ligand_target_matrix, ligands_position):\n",
      "“all target gene probability score predictions have same value”\n",
      "Warning message in cor(prediction, response):\n",
      "“the standard deviation is zero”\n",
      "Warning message in cor(prediction, response, method = \"s\"):\n",
      "“the standard deviation is zero”\n",
      "Warning message in evaluate_target_prediction(setting, ligand_target_matrix, ligands_position):\n",
      "“all target gene probability score predictions have same value”\n",
      "Warning message in cor(prediction, response):\n",
      "“the standard deviation is zero”\n",
      "Warning message in cor(prediction, response, method = \"s\"):\n",
      "“the standard deviation is zero”\n",
      "Warning message in evaluate_target_prediction(setting, ligand_target_matrix, ligands_position):\n",
      "“all target gene probability score predictions have same value”\n",
      "Warning message in cor(prediction, response):\n",
      "“the standard deviation is zero”\n",
      "Warning message in cor(prediction, response, method = \"s\"):\n",
      "“the standard deviation is zero”\n",
      "Warning message in evaluate_target_prediction(setting, ligand_target_matrix, ligands_position):\n",
      "“all target gene probability score predictions have same value”\n",
      "Warning message in cor(prediction, response):\n",
      "“the standard deviation is zero”\n",
      "Warning message in cor(prediction, response, method = \"s\"):\n",
      "“the standard deviation is zero”\n",
      "Warning message in evaluate_target_prediction(setting, ligand_target_matrix, ligands_position):\n",
      "“all target gene probability score predictions have same value”\n",
      "Warning message in cor(prediction, response):\n",
      "“the standard deviation is zero”\n",
      "Warning message in cor(prediction, response, method = \"s\"):\n",
      "“the standard deviation is zero”\n",
      "Warning message in evaluate_target_prediction(setting, ligand_target_matrix, ligands_position):\n",
      "“all target gene probability score predictions have same value”\n",
      "Warning message in cor(prediction, response):\n",
      "“the standard deviation is zero”\n",
      "Warning message in cor(prediction, response, method = \"s\"):\n",
      "“the standard deviation is zero”\n",
      "Warning message in evaluate_target_prediction(setting, ligand_target_matrix, ligands_position):\n",
      "“all target gene probability score predictions have same value”\n",
      "Warning message in cor(prediction, response):\n",
      "“the standard deviation is zero”\n",
      "Warning message in cor(prediction, response, method = \"s\"):\n",
      "“the standard deviation is zero”\n",
      "Warning message in evaluate_target_prediction(setting, ligand_target_matrix, ligands_position):\n",
      "“all target gene probability score predictions have same value”\n",
      "Warning message in cor(prediction, response):\n",
      "“the standard deviation is zero”\n",
      "Warning message in cor(prediction, response, method = \"s\"):\n",
      "“the standard deviation is zero”\n",
      "Warning message in evaluate_target_prediction(setting, ligand_target_matrix, ligands_position):\n",
      "“all target gene probability score predictions have same value”\n",
      "Warning message in cor(prediction, response):\n",
      "“the standard deviation is zero”\n",
      "Warning message in cor(prediction, response, method = \"s\"):\n",
      "“the standard deviation is zero”\n",
      "Warning message in evaluate_target_prediction(setting, ligand_target_matrix, ligands_position):\n",
      "“all target gene probability score predictions have same value”\n",
      "Warning message in cor(prediction, response):\n",
      "“the standard deviation is zero”\n",
      "Warning message in cor(prediction, response, method = \"s\"):\n",
      "“the standard deviation is zero”\n",
      "Warning message in evaluate_target_prediction(setting, ligand_target_matrix, ligands_position):\n",
      "“all target gene probability score predictions have same value”\n",
      "Warning message in cor(prediction, response):\n",
      "“the standard deviation is zero”\n",
      "Warning message in cor(prediction, response, method = \"s\"):\n",
      "“the standard deviation is zero”\n",
      "Warning message in evaluate_target_prediction(setting, ligand_target_matrix, ligands_position):\n",
      "“all target gene probability score predictions have same value”\n",
      "Warning message in cor(prediction, response):\n",
      "“the standard deviation is zero”\n",
      "Warning message in cor(prediction, response, method = \"s\"):\n",
      "“the standard deviation is zero”\n",
      "Warning message in evaluate_target_prediction(setting, ligand_target_matrix, ligands_position):\n",
      "“all target gene probability score predictions have same value”\n",
      "Warning message in cor(prediction, response):\n",
      "“the standard deviation is zero”\n",
      "Warning message in cor(prediction, response, method = \"s\"):\n",
      "“the standard deviation is zero”\n",
      "Warning message in evaluate_target_prediction(setting, ligand_target_matrix, ligands_position):\n",
      "“all target gene probability score predictions have same value”\n",
      "Warning message in cor(prediction, response):\n",
      "“the standard deviation is zero”\n",
      "Warning message in cor(prediction, response, method = \"s\"):\n",
      "“the standard deviation is zero”\n",
      "Warning message in evaluate_target_prediction(setting, ligand_target_matrix, ligands_position):\n",
      "“all target gene probability score predictions have same value”\n",
      "Warning message in cor(prediction, response):\n",
      "“the standard deviation is zero”\n",
      "Warning message in cor(prediction, response, method = \"s\"):\n",
      "“the standard deviation is zero”\n",
      "Warning message in evaluate_target_prediction(setting, ligand_target_matrix, ligands_position):\n",
      "“all target gene probability score predictions have same value”\n",
      "Warning message in cor(prediction, response):\n",
      "“the standard deviation is zero”\n",
      "Warning message in cor(prediction, response, method = \"s\"):\n",
      "“the standard deviation is zero”\n",
      "Warning message in evaluate_target_prediction(setting, ligand_target_matrix, ligands_position):\n",
      "“all target gene probability score predictions have same value”\n",
      "Warning message in cor(prediction, response):\n",
      "“the standard deviation is zero”\n",
      "Warning message in cor(prediction, response, method = \"s\"):\n",
      "“the standard deviation is zero”\n",
      "Warning message in evaluate_target_prediction(setting, ligand_target_matrix, ligands_position):\n",
      "“all target gene probability score predictions have same value”\n",
      "Warning message in cor(prediction, response):\n",
      "“the standard deviation is zero”\n",
      "Warning message in cor(prediction, response, method = \"s\"):\n",
      "“the standard deviation is zero”\n",
      "Warning message in evaluate_target_prediction(setting, ligand_target_matrix, ligands_position):\n",
      "“all target gene probability score predictions have same value”\n",
      "Warning message in cor(prediction, response):\n",
      "“the standard deviation is zero”\n",
      "Warning message in cor(prediction, response, method = \"s\"):\n",
      "“the standard deviation is zero”\n",
      "Warning message in evaluate_target_prediction(setting, ligand_target_matrix, ligands_position):\n",
      "“all target gene probability score predictions have same value”\n",
      "Warning message in cor(prediction, response):\n",
      "“the standard deviation is zero”\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in cor(prediction, response, method = \"s\"):\n",
      "“the standard deviation is zero”\n",
      "Warning message in evaluate_target_prediction(setting, ligand_target_matrix, ligands_position):\n",
      "“all target gene probability score predictions have same value”\n",
      "Warning message in cor(prediction, response):\n",
      "“the standard deviation is zero”\n",
      "Warning message in cor(prediction, response, method = \"s\"):\n",
      "“the standard deviation is zero”\n",
      "Warning message in evaluate_target_prediction(setting, ligand_target_matrix, ligands_position):\n",
      "“all target gene probability score predictions have same value”\n",
      "Warning message in cor(prediction, response):\n",
      "“the standard deviation is zero”\n",
      "Warning message in cor(prediction, response, method = \"s\"):\n",
      "“the standard deviation is zero”\n",
      "Warning message in evaluate_target_prediction(setting, ligand_target_matrix, ligands_position):\n",
      "“all target gene probability score predictions have same value”\n",
      "Warning message in cor(prediction, response):\n",
      "“the standard deviation is zero”\n",
      "Warning message in cor(prediction, response, method = \"s\"):\n",
      "“the standard deviation is zero”\n",
      "Warning message in evaluate_target_prediction(setting, ligand_target_matrix, ligands_position):\n",
      "“all target gene probability score predictions have same value”\n",
      "Warning message in cor(prediction, response):\n",
      "“the standard deviation is zero”\n",
      "Warning message in cor(prediction, response, method = \"s\"):\n",
      "“the standard deviation is zero”\n",
      "Warning message in evaluate_target_prediction(setting, ligand_target_matrix, ligands_position):\n",
      "“all target gene probability score predictions have same value”\n",
      "Warning message in cor(prediction, response):\n",
      "“the standard deviation is zero”\n",
      "Warning message in cor(prediction, response, method = \"s\"):\n",
      "“the standard deviation is zero”\n",
      "Warning message in evaluate_target_prediction(setting, ligand_target_matrix, ligands_position):\n",
      "“all target gene probability score predictions have same value”\n",
      "Warning message in cor(prediction, response):\n",
      "“the standard deviation is zero”\n",
      "Warning message in cor(prediction, response, method = \"s\"):\n",
      "“the standard deviation is zero”\n",
      "Warning message in evaluate_target_prediction(setting, ligand_target_matrix, ligands_position):\n",
      "“all target gene probability score predictions have same value”\n",
      "Warning message in cor(prediction, response):\n",
      "“the standard deviation is zero”\n",
      "Warning message in cor(prediction, response, method = \"s\"):\n",
      "“the standard deviation is zero”\n",
      "Warning message in evaluate_target_prediction(setting, ligand_target_matrix, ligands_position):\n",
      "“all target gene probability score predictions have same value”\n",
      "Warning message in cor(prediction, response):\n",
      "“the standard deviation is zero”\n",
      "Warning message in cor(prediction, response, method = \"s\"):\n",
      "“the standard deviation is zero”\n",
      "Warning message in evaluate_target_prediction(setting, ligand_target_matrix, ligands_position):\n",
      "“all target gene probability score predictions have same value”\n",
      "Warning message in cor(prediction, response):\n",
      "“the standard deviation is zero”\n",
      "Warning message in cor(prediction, response, method = \"s\"):\n",
      "“the standard deviation is zero”\n",
      "Warning message in evaluate_target_prediction(setting, ligand_target_matrix, ligands_position):\n",
      "“all target gene probability score predictions have same value”\n",
      "Warning message in cor(prediction, response):\n",
      "“the standard deviation is zero”\n",
      "Warning message in cor(prediction, response, method = \"s\"):\n",
      "“the standard deviation is zero”\n",
      "Warning message in evaluate_target_prediction(setting, ligand_target_matrix, ligands_position):\n",
      "“all target gene probability score predictions have same value”\n",
      "Warning message in cor(prediction, response):\n",
      "“the standard deviation is zero”\n",
      "Warning message in cor(prediction, response, method = \"s\"):\n",
      "“the standard deviation is zero”\n",
      "Warning message in evaluate_target_prediction(setting, ligand_target_matrix, ligands_position):\n",
      "“all target gene probability score predictions have same value”\n",
      "Warning message in cor(prediction, response):\n",
      "“the standard deviation is zero”\n",
      "Warning message in cor(prediction, response, method = \"s\"):\n",
      "“the standard deviation is zero”\n",
      "Warning message in evaluate_target_prediction(setting, ligand_target_matrix, ligands_position):\n",
      "“all target gene probability score predictions have same value”\n",
      "Warning message in cor(prediction, response):\n",
      "“the standard deviation is zero”\n",
      "Warning message in cor(prediction, response, method = \"s\"):\n",
      "“the standard deviation is zero”\n",
      "Warning message in evaluate_target_prediction(setting, ligand_target_matrix, ligands_position):\n",
      "“all target gene probability score predictions have same value”\n",
      "Warning message in cor(prediction, response):\n",
      "“the standard deviation is zero”\n",
      "Warning message in cor(prediction, response, method = \"s\"):\n",
      "“the standard deviation is zero”\n",
      "Warning message in evaluate_target_prediction(setting, ligand_target_matrix, ligands_position):\n",
      "“all target gene probability score predictions have same value”\n",
      "Warning message in cor(prediction, response):\n",
      "“the standard deviation is zero”\n",
      "Warning message in cor(prediction, response, method = \"s\"):\n",
      "“the standard deviation is zero”\n",
      "Warning message in evaluate_target_prediction(setting, ligand_target_matrix, ligands_position):\n",
      "“all target gene probability score predictions have same value”\n",
      "Warning message in cor(prediction, response):\n",
      "“the standard deviation is zero”\n",
      "Warning message in cor(prediction, response, method = \"s\"):\n",
      "“the standard deviation is zero”\n",
      "Warning message in evaluate_target_prediction(setting, ligand_target_matrix, ligands_position):\n",
      "“all target gene probability score predictions have same value”\n",
      "Warning message in cor(prediction, response):\n",
      "“the standard deviation is zero”\n",
      "Warning message in cor(prediction, response, method = \"s\"):\n",
      "“the standard deviation is zero”\n",
      "Warning message in evaluate_target_prediction(setting, ligand_target_matrix, ligands_position):\n",
      "“all target gene probability score predictions have same value”\n",
      "Warning message in cor(prediction, response):\n",
      "“the standard deviation is zero”\n",
      "Warning message in cor(prediction, response, method = \"s\"):\n",
      "“the standard deviation is zero”\n"
     ]
    }
   ],
   "source": [
    "# Get unique ligands and receptors from lr_network\n",
    "\n",
    "ligands_in_db = lr_network %>% pull(from) %>% unique()\n",
    "receptors_in_db = lr_network %>% pull(to) %>% unique()\n",
    "\n",
    "# Initialize an empty list to store results\n",
    "results=list()\n",
    "ligand_activities_list <- list()\n",
    "\n",
    "for (receiver in cell_type_list){\n",
    "    \n",
    "    # Get DE genes for the receiver cell type\n",
    "    geneset_oi <- DE_tables[[receiver]] %>% filter(p_val_adj <= 0.05 & abs(avg_log2FC) >= 0.25) %>% pull(gene)\n",
    "    \n",
    "#     n_DE_genes <- length(geneset_oi)\n",
    "    \n",
    "    # Iterate over each sample\n",
    "    for (sample in names(expressed_genes_list)){\n",
    "        \n",
    "        # Get expressed genes for the receiver cell type in the current sample\n",
    "        expressed_genes_receiver <- expressed_genes_list[[sample]][[receiver]]\n",
    "        \n",
    "        # Get the intersection of DE genes and expressed genes for the receiver cell type in the current sample\n",
    "        geneset_oi_sample_expressed <- geneset_oi[geneset_oi %in% expressed_genes_receiver]\n",
    "        \n",
    "        # Get the intersection of DE genes and ligand-target gene matrix rows\n",
    "        geneset_oi_sample_expressed <- geneset_oi_sample_expressed[geneset_oi_sample_expressed %in% rownames(ligand_target_matrix)]\n",
    "        \n",
    "        # Get the intersection of expressed genes for the receiver cell type in the current sample and ligand-target gene matrix rows\n",
    "        background_expressed_genes  <- expressed_genes_receiver[expressed_genes_receiver %in% rownames(ligand_target_matrix)]\n",
    "        \n",
    "        # Iterate over each sender cell type\n",
    "        for (sender in names(expressed_genes_list[[sample]])){\n",
    "            \n",
    "            # Get expressed genes for the sender cell type in the current sample\n",
    "            expressed_genes_sender <- expressed_genes_list[[sample]][[sender]]\n",
    "            \n",
    "            # Check if the sender cell type is not null for the receiver cell type\n",
    "            if (!is.null(expressed_genes_list[[sample]][[receiver]])){\n",
    "                \n",
    "                # Get the intersection of expressed ligands and ligands in the lr_network\n",
    "                expressed_ligands <- expressed_genes_sender[expressed_genes_sender %in% ligands_in_db]\n",
    "                expressed_receptors <- expressed_genes_receiver[expressed_genes_receiver %in% ligands_in_db]\n",
    "                \n",
    "                # Get the potential ligands based on the expressed ligands and receptors\n",
    "                potential_ligands = lr_network %>% \n",
    "                    filter(from %in% expressed_ligands & to %in% expressed_receptors) %>% \n",
    "                        pull(from) %>% unique()\n",
    "                \n",
    "                \n",
    "                # Predict ligand activities with NicheNet\n",
    "                ligand_activities = predict_ligand_activities(geneset = geneset_oi_sample_expressed, \n",
    "                                                              background_expressed_genes = background_expressed_genes, \n",
    "                                                              ligand_target_matrix = ligand_target_matrix, \n",
    "                                                              potential_ligands = potential_ligands)\n",
    "                \n",
    "                # Arrange ligand activities by pearson correlation coefficient and add a rank column\n",
    "                ligand_activities = ligand_activities %>% arrange(-pearson) %>% mutate(rank = rank(desc(pearson)))\n",
    "                \n",
    "                \n",
    "                # Get the potential ligands based on the ligand activities\n",
    "                potential_ligands = ligand_activities %>% pull(test_ligand) %>% unique()\n",
    "                \n",
    "                # Filter the lr_network to only include interactions between potential ligands and expressed receptors\n",
    "                lr_network_expressed = lr_network %>% \n",
    "                    filter(from %in% potential_ligands & to %in% expressed_receptors) %>% distinct(from,to)\n",
    "                \n",
    "                \n",
    "                # Get the corresponding interactions from the weighted network\n",
    "                lr_network_df_large = weighted_networks_lr %>% filter(from %in% potential_ligands & \n",
    "                                                                      to %in% expressed_receptors)\n",
    "                \n",
    "                # Add metadata to the network dataframe\n",
    "                lr_network_df_large$sample <- sample\n",
    "                lr_network_df_large$cell_type_l <- sender\n",
    "                lr_network_df_large$cell_type_r <- receiver\n",
    "                lr_network_df_large$pair <- paste(lr_network_df_large$from,lr_network_df_large$to, sep=\"_\")\n",
    "                lr_network_df_large$interaction = paste0(lr_network_df_large$cell_type_l,\":\",\n",
    "                                                         lr_network_df_large$from,\"_\",\n",
    "                                                         lr_network_df_large$cell_type_r,\":\",\n",
    "                                                         lr_network_df_large$to)\n",
    "                \n",
    "                \n",
    "                df_new <- transform_interaction_df(lr_network_df_large)\n",
    "                results[[sample]] <- rbind(results[[sample]],df_new)\n",
    "                ligand_activities_list[[sample]][[sender]][[receiver]] <- ligand_activities\n",
    "                \n",
    "#                 print(receiver)\n",
    "#                 print(nrow(lr_network_df_large))\n",
    "#                 write.csv(ligand_activities, paste(\"results6/\",sample,\"_\",sender,\"_\",receiver,\"_ligand_activities.csv\", sep=\"\"))\n",
    "#                 write.csv(lr_network_df_large, paste(\"results6/\",sample,\"_\",sender,\"_\",receiver,\"_LR.csv\", sep=\"\"))\n",
    "            }\n",
    "        }\n",
    "        \n",
    "    }\n",
    "    \n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "296cb963",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define a variable called `result` that will hold the output of the Reduce function\n",
    "matrix_result <- Reduce(\n",
    "  \n",
    "  # The `Reduce()` function takes two arguments: a function and a list.\n",
    "  # In this case, the function is an anonymous function defined using the `function()` keyword.\n",
    "  # This function takes two arguments `x` and `y` and performs a full join between them using the `full_join()` function from the `dplyr` package.\n",
    "  # The `by = \"interaction\"` argument specifies that the join should be performed on the \"interaction\" column.\n",
    "  function(x, y) full_join(x, y, by = \"interaction\"), \n",
    "  \n",
    "  # The second argument to the `Reduce()` function is a list called `results`.\n",
    "  # This list contains data frames that need to be joined together.\n",
    "  results\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "7f170873",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A tibble: 6 × 14</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>interaction</th><th scope=col>AML-0024</th><th scope=col>AML-0160</th><th scope=col>AML-0693</th><th scope=col>AML-1371</th><th scope=col>AML-2123</th><th scope=col>AML-3133</th><th scope=col>AML-4340</th><th scope=col>healthy-1</th><th scope=col>healthy-2</th><th scope=col>healthy-3</th><th scope=col>healthy-4</th><th scope=col>healthy-4003</th><th scope=col>healthy-5</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>Mono:ADAM10_Mono:CADM1</td><td>0.19214601</td><td>        NA</td><td>        NA</td><td>0.19214601</td><td>        NA</td><td>0.19214601</td><td>        NA</td><td>        NA</td><td>        NA</td><td>        NA</td><td>        NA</td><td>        NA</td><td>        NA</td></tr>\n",
       "\t<tr><td>Mono:ADAM10_Mono:CD44 </td><td>0.09211742</td><td>0.09211742</td><td>0.09211742</td><td>0.09211742</td><td>0.09211742</td><td>0.09211742</td><td>0.09211742</td><td>0.09211742</td><td>0.09211742</td><td>0.09211742</td><td>0.09211742</td><td>0.09211742</td><td>0.09211742</td></tr>\n",
       "\t<tr><td>Mono:ADAM10_Mono:IL6R </td><td>0.09416423</td><td>0.09416423</td><td>0.09416423</td><td>0.09416423</td><td>0.09416423</td><td>0.09416423</td><td>0.09416423</td><td>0.09416423</td><td>0.09416423</td><td>0.09416423</td><td>0.09416423</td><td>0.09416423</td><td>0.09416423</td></tr>\n",
       "\t<tr><td>Mono:ADAM9_Mono:ITGB1 </td><td>0.36129136</td><td>0.36129136</td><td>0.36129136</td><td>        NA</td><td>0.36129136</td><td>0.36129136</td><td>0.36129136</td><td>0.36129136</td><td>0.36129136</td><td>0.36129136</td><td>0.36129136</td><td>        NA</td><td>        NA</td></tr>\n",
       "\t<tr><td>Mono:ADM_Mono:ACKR3   </td><td>0.27477269</td><td>        NA</td><td>        NA</td><td>        NA</td><td>        NA</td><td>        NA</td><td>        NA</td><td>        NA</td><td>        NA</td><td>        NA</td><td>        NA</td><td>        NA</td><td>        NA</td></tr>\n",
       "\t<tr><td>Mono:ANG_Mono:PLXNB2  </td><td>0.10136751</td><td>        NA</td><td>        NA</td><td>        NA</td><td>        NA</td><td>        NA</td><td>        NA</td><td>        NA</td><td>        NA</td><td>        NA</td><td>        NA</td><td>        NA</td><td>        NA</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A tibble: 6 × 14\n",
       "\\begin{tabular}{llllllllllllll}\n",
       " interaction & AML-0024 & AML-0160 & AML-0693 & AML-1371 & AML-2123 & AML-3133 & AML-4340 & healthy-1 & healthy-2 & healthy-3 & healthy-4 & healthy-4003 & healthy-5\\\\\n",
       " <chr> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t Mono:ADAM10\\_Mono:CADM1 & 0.19214601 &         NA &         NA & 0.19214601 &         NA & 0.19214601 &         NA &         NA &         NA &         NA &         NA &         NA &         NA\\\\\n",
       "\t Mono:ADAM10\\_Mono:CD44  & 0.09211742 & 0.09211742 & 0.09211742 & 0.09211742 & 0.09211742 & 0.09211742 & 0.09211742 & 0.09211742 & 0.09211742 & 0.09211742 & 0.09211742 & 0.09211742 & 0.09211742\\\\\n",
       "\t Mono:ADAM10\\_Mono:IL6R  & 0.09416423 & 0.09416423 & 0.09416423 & 0.09416423 & 0.09416423 & 0.09416423 & 0.09416423 & 0.09416423 & 0.09416423 & 0.09416423 & 0.09416423 & 0.09416423 & 0.09416423\\\\\n",
       "\t Mono:ADAM9\\_Mono:ITGB1  & 0.36129136 & 0.36129136 & 0.36129136 &         NA & 0.36129136 & 0.36129136 & 0.36129136 & 0.36129136 & 0.36129136 & 0.36129136 & 0.36129136 &         NA &         NA\\\\\n",
       "\t Mono:ADM\\_Mono:ACKR3    & 0.27477269 &         NA &         NA &         NA &         NA &         NA &         NA &         NA &         NA &         NA &         NA &         NA &         NA\\\\\n",
       "\t Mono:ANG\\_Mono:PLXNB2   & 0.10136751 &         NA &         NA &         NA &         NA &         NA &         NA &         NA &         NA &         NA &         NA &         NA &         NA\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A tibble: 6 × 14\n",
       "\n",
       "| interaction &lt;chr&gt; | AML-0024 &lt;dbl&gt; | AML-0160 &lt;dbl&gt; | AML-0693 &lt;dbl&gt; | AML-1371 &lt;dbl&gt; | AML-2123 &lt;dbl&gt; | AML-3133 &lt;dbl&gt; | AML-4340 &lt;dbl&gt; | healthy-1 &lt;dbl&gt; | healthy-2 &lt;dbl&gt; | healthy-3 &lt;dbl&gt; | healthy-4 &lt;dbl&gt; | healthy-4003 &lt;dbl&gt; | healthy-5 &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| Mono:ADAM10_Mono:CADM1 | 0.19214601 |         NA |         NA | 0.19214601 |         NA | 0.19214601 |         NA |         NA |         NA |         NA |         NA |         NA |         NA |\n",
       "| Mono:ADAM10_Mono:CD44  | 0.09211742 | 0.09211742 | 0.09211742 | 0.09211742 | 0.09211742 | 0.09211742 | 0.09211742 | 0.09211742 | 0.09211742 | 0.09211742 | 0.09211742 | 0.09211742 | 0.09211742 |\n",
       "| Mono:ADAM10_Mono:IL6R  | 0.09416423 | 0.09416423 | 0.09416423 | 0.09416423 | 0.09416423 | 0.09416423 | 0.09416423 | 0.09416423 | 0.09416423 | 0.09416423 | 0.09416423 | 0.09416423 | 0.09416423 |\n",
       "| Mono:ADAM9_Mono:ITGB1  | 0.36129136 | 0.36129136 | 0.36129136 |         NA | 0.36129136 | 0.36129136 | 0.36129136 | 0.36129136 | 0.36129136 | 0.36129136 | 0.36129136 |         NA |         NA |\n",
       "| Mono:ADM_Mono:ACKR3    | 0.27477269 |         NA |         NA |         NA |         NA |         NA |         NA |         NA |         NA |         NA |         NA |         NA |         NA |\n",
       "| Mono:ANG_Mono:PLXNB2   | 0.10136751 |         NA |         NA |         NA |         NA |         NA |         NA |         NA |         NA |         NA |         NA |         NA |         NA |\n",
       "\n"
      ],
      "text/plain": [
       "  interaction            AML-0024   AML-0160   AML-0693   AML-1371   AML-2123  \n",
       "1 Mono:ADAM10_Mono:CADM1 0.19214601         NA         NA 0.19214601         NA\n",
       "2 Mono:ADAM10_Mono:CD44  0.09211742 0.09211742 0.09211742 0.09211742 0.09211742\n",
       "3 Mono:ADAM10_Mono:IL6R  0.09416423 0.09416423 0.09416423 0.09416423 0.09416423\n",
       "4 Mono:ADAM9_Mono:ITGB1  0.36129136 0.36129136 0.36129136         NA 0.36129136\n",
       "5 Mono:ADM_Mono:ACKR3    0.27477269         NA         NA         NA         NA\n",
       "6 Mono:ANG_Mono:PLXNB2   0.10136751         NA         NA         NA         NA\n",
       "  AML-3133   AML-4340   healthy-1  healthy-2  healthy-3  healthy-4 \n",
       "1 0.19214601         NA         NA         NA         NA         NA\n",
       "2 0.09211742 0.09211742 0.09211742 0.09211742 0.09211742 0.09211742\n",
       "3 0.09416423 0.09416423 0.09416423 0.09416423 0.09416423 0.09416423\n",
       "4 0.36129136 0.36129136 0.36129136 0.36129136 0.36129136 0.36129136\n",
       "5         NA         NA         NA         NA         NA         NA\n",
       "6         NA         NA         NA         NA         NA         NA\n",
       "  healthy-4003 healthy-5 \n",
       "1         NA           NA\n",
       "2 0.09211742   0.09211742\n",
       "3 0.09416423   0.09416423\n",
       "4         NA           NA\n",
       "5         NA           NA\n",
       "6         NA           NA"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(matrix_result)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "5b7038f3",
   "metadata": {},
   "outputs": [],
   "source": [
    "write.csv(matrix_result, paste0(output_dir,\"NicheNet_results.csv\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a649db8a",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R [conda env:nnet]",
   "language": "R",
   "name": "conda-env-nnet-r"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.2.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     