{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "490cf646",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "========================================\n",
      "ComplexHeatmap version 2.10.0\n",
      "Bioconductor page: http://bioconductor.org/packages/ComplexHeatmap/\n",
      "Github page: https://github.com/jokergoo/ComplexHeatmap\n",
      "Documentation: http://jokergoo.github.io/ComplexHeatmap-reference\n",
      "\n",
      "If you use it in published research, please cite:\n",
      "Gu, Z. Complex heatmaps reveal patterns and correlations in multidimensional \n",
      "  genomic data. Bioinformatics 2016.\n",
      "\n",
      "The new InteractiveComplexHeatmap package can directly export static \n",
      "complex heatmaps into an interactive Shiny app with zero effort. Have a try!\n",
      "\n",
      "This message can be suppressed by:\n",
      "  suppressPackageStartupMessages(library(ComplexHeatmap))\n",
      "========================================\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘igraph’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    decompose, spectrum\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:base’:\n",
      "\n",
      "    union\n",
      "\n",
      "\n",
      "Loading required package: circlize\n",
      "\n",
      "========================================\n",
      "circlize version 0.4.15\n",
      "CRAN page: https://cran.r-project.org/package=circlize\n",
      "Github page: https://github.com/jokergoo/circlize\n",
      "Documentation: https://jokergoo.github.io/circlize_book/book/\n",
      "\n",
      "If you use it in published research, please cite:\n",
      "Gu, Z. circlize implements and enhances circular visualization\n",
      "  in R. Bioinformatics 2014.\n",
      "\n",
      "This message can be suppressed by:\n",
      "  suppressPackageStartupMessages(library(circlize))\n",
      "========================================\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘circlize’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:igraph’:\n",
      "\n",
      "    degree\n",
      "\n",
      "\n",
      "Loading required package: R.oo\n",
      "\n",
      "Loading required package: R.methodsS3\n",
      "\n",
      "R.methodsS3 v1.8.2 (2022-06-13 22:00:14 UTC) successfully loaded. See ?R.methodsS3 for help.\n",
      "\n",
      "R.oo v1.25.0 (2022-06-12 02:20:02 UTC) successfully loaded. See ?R.oo for help.\n",
      "\n",
      "\n",
      "Attaching package: ‘R.oo’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:R.methodsS3’:\n",
      "\n",
      "    throw\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:igraph’:\n",
      "\n",
      "    hierarchy\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:methods’:\n",
      "\n",
      "    getClasses, getMethods\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    attach, detach, load, save\n",
      "\n",
      "\n",
      "R.utils v2.12.2 (2022-11-11 22:00:03 UTC) successfully loaded. See ?R.utils for help.\n",
      "\n",
      "\n",
      "Attaching package: ‘R.utils’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:ComplexHeatmap’:\n",
      "\n",
      "    draw\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:utils’:\n",
      "\n",
      "    timestamp\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    cat, commandArgs, getOption, isOpen, nullfile, parse, warnings\n",
      "\n",
      "\n",
      "Attaching SeuratObject\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# libraries\n",
    "library(community)\n",
    "library(ggplot2)\n",
    "library(gridExtra)\n",
    "library(grid)\n",
    "library(ComplexHeatmap)\n",
    "library(dendsort)\n",
    "library(igraph)\n",
    "require(circlize)\n",
    "library(R.utils)\n",
    "library(data.table) #to read gz file\n",
    "library(Seurat)\n",
    "library(nichenetr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "e05fd56d",
   "metadata": {},
   "outputs": [],
   "source": [
    "input_dir <- \"../../../../results/data_preprocessing/Lasry/preprocessed/\"\n",
    "output_dir <- \"../../../../results/method_comparison/compare_algorithms/community/\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "b40a8483",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"load counts\"\n",
      "'data.frame':\t15770 obs. of  46702 variables:\n",
      " $ 2020.09.15.AML0024.CATCAAGGTTAGCGGA           : num  0 0 0 0 0 0 0 0 0 0 ...\n",
      " $ 2020.09.15.AML0024.CATCAAGTCCGAGAAG           : num  0 0 0 0 0 0 0 0 0 0 ...\n",
      " $ 2020.09.15.AML0024.CATCCACAGGGACCAT           : num  0 0 0 0 0 ...\n",
      " $ 2020.09.15.AML0024.CCTCAACAGAGCAAGA           : num  0 0 0 0 0 0 0 0 0 0 ...\n",
      " $ 2020.09.15.AML0024.CCTCAACAGTTCCATG           : num  0 0.69 0 0 0 ...\n",
      " $ 2020.09.15.AML0024.CCTCAACGTAGAATAC           : num  0 0 0 0 0 0 0 0 0 0 ...\n",
      " $ 2020.09.15.AML0024.CCTCAACGTTCTCCCA           : num  0 0 0 0 0 0 0 0 0 0 ...\n",
      " $ 2020.09.15.AML0024.CCTCAACTCCGAACGC           : num  0 0 0 0 0 ...\n",
      " $ 2020.09.15.AML0024.CCTCAACTCTAGTCAG           : num  0 0 0 0 0 0 0 0 0 0 ...\n",
      " $ 2020.09.15.AML0024.CCTCACAAGACAGTCG           : num  0 0 0 0 0 0 0 0 0 0 ...\n",
      " $ 2020.09.15.AML0024.CCTCACACAATTGCCA           : num  0 0 0 0 0 0 0 0 0 0 ...\n",
      " $ 2020.09.15.AML0024.CCTCACACAGAACTAA           : num  0 0 0 0 0 0 0 0 0 0 ...\n",
      " $ 2020.09.15.AML0024.CCTCACACAGCTGAAG           : num  0 0 0 0 0 0 0 0 0 0 ...\n",
      " $ 2020.09.15.AML0024.CCTCACAGTAATCAGA           : num  0 0 0 0 0 ...\n",
      " $ 2020.09.15.AML0024.CCTCACAGTACTCCGG           : num  0 0 0 0 0 0 0 0 0 0 ...\n",
      " $ 2020.09.15.AML0024.CCTCAACAGAACGCGT           : num  0 0 0.732 0 0 ...\n",
      " $ 2020.09.15.AML0024.CATACTTAGGCGTTAG           : num  0 0 0 0 0 0 0 0 0 0 ...\n",
      " $ 2020.09.15.AML0024.CATACTTAGTTGCATC           : num  0 0 0 0 0 ...\n",
      " $ 2020.09.15.AML0024.CATACTTCAACTTGGT           : num  0 0 0 0 0 0 0 0 0 0 ...\n",
      " $ 2020.09.15.AML0024.CATACTTTCCTTGACC           : num  0 0 0 0 0 0 0 0 0 0 ...\n",
      " $ 2020.09.15.AML0024.CATAGACAGCACTAAA           : num  0 0 0 0 0 0 0 0 0 0 ...\n",
      " $ 2020.09.15.AML0024.CATAGACAGGATCACG           : num  0 0 0 0 0 0 0 0 0 0 ...\n",
      " $ 2020.09.15.AML0024.CATAGACAGGGAGTGG           : num  0 0 0 0 0 0 0 0 0 0 ...\n",
      " $ 2020.09.15.AML0024.CATAGACCACACAGCC           : num  0 0 0 0 0 ...\n",
      " $ 2020.09.15.AML0024.CATAGACTCCATTGGA           : num  0 0 0 0 0 0 0 0 0 0 ...\n",
      " $ 2020.09.15.AML0024.CATCAAGAGAGTTCGG           : num  0 0 0 0 0 0 0 0 0 0 ...\n",
      " $ 2020.09.15.AML0024.CATCAAGAGAGTTGCG           : num  0 0 0 0 0 0 0 0 0 0 ...\n",
      " $ 2020.09.15.AML0024.CCTAAGATCGAGAGCA           : num  0 0 0 0 0 0 0 0 0 0 ...\n",
      " $ 2020.09.15.AML0024.CCTACGTAGCCGTAAG           : num  0 0 0 0 0 0 0 0 0 0 ...\n",
      " $ 2020.09.15.AML0024.CCTACGTCATCCTTCG           : num  0 0 0 0 0.452 ...\n",
      " $ 2020.09.15.AML0024.CCTACGTGTATTTCTC           : num  0 0 0 0 0 0 0 0 0 0 ...\n",
      " $ 2020.09.15.AML0024.CCTACGTGTCTCAGAT           : num  0 0 0 0 0 ...\n",
      " $ 2020.09.15.AML0024.CCTACGTTCCACAAGT           : num  0 0 0 0 0 0 0 0 0 0 ...\n",
      " $ 2020.09.15.AML0024.CCTATCGAGAATCCCT           : num  0 0 0 0 0 0 0 0 0 0 ...\n",
      " $ 2020.09.15.AML0024.CCTATCGAGGAGACCT           : num  0 0 0 0 0 0 0 0 0 0 ...\n",
      " $ 2020.09.15.AML0024.CCTATCGAGTCGAGGT           : num  0 0 0 0 0 0 0 0 0 0 ...\n",
      " $ 2020.09.15.AML0024.CCTATCGCAGAACGCA           : num  0 0 0 0 0 ...\n",
      " $ 2020.09.15.AML0024.CCTATCGCAGCTCGGT           : num  0 0 0 0 0 0 0 0 0 0 ...\n",
      " $ 2020.09.15.AML0024.CCTATCGGTAAGTTGA           : num  0 0 0 0 0 0 0 0 0 0 ...\n",
      " $ 2020.09.15.AML0024.CCTATCGGTCAGTCCG           : num  0 0 0 0 0 0 0 0 0 0 ...\n",
      " $ 2020.09.15.AML0024.CCTATCGTCCGTGTAA           : num  0 0 0 0 0 0 0 0 0 0 ...\n",
      " $ 2020.09.15.AML0024.CCTATCGTCGTAGCCG           : num  0 0 0 0 0 0 0 0 0 0 ...\n",
      " $ 2020.09.15.AML0024.CCTATCGTCTACCACC           : num  0 0 0.491 0 0 ...\n",
      " $ 2020.09.15.AML0024.AAACCCACAACATACC           : num  0 0 0 0 0 0 0 0 0 0 ...\n",
      " $ 2020.09.15.AML0024.AAACCCAGTACAACGG           : num  0 0 0 0 0 0 0 0 0 0 ...\n",
      " $ 2020.09.15.AML0024.AAACCCAGTTGTAGCT           : num  0 0 0 0 0 0 0 0 0 0 ...\n",
      " $ 2020.09.15.AML0024.AAACCCATCATCTGTT           : num  0 0 0 0 0 ...\n",
      " $ 2020.09.15.AML0024.AAACCCATCTCAGAAC           : num  0 0 0 0 0 ...\n",
      " $ 2020.09.15.AML0024.AAACGAACAAGTTCGT           : num  0 0 0 0 0 0 0 0 0 0 ...\n",
      " $ 2020.09.15.AML0024.AAACGAACAGCAGTTT           : num  0 0 0 0 0 0 0 0 0 0 ...\n",
      " $ 2020.09.15.AML0024.AAACGAAGTAACAGTA           : num  0 0 0 0 0 0 0 0 0 0 ...\n",
      " $ 2020.09.15.AML0024.AAACGAATCTTTGGAG           : num  0 0 0 0 0 0 0 0 0 0 ...\n",
      " $ 2020.09.15.AML0024.AAACGCTAGATTAGTG           : num  0 0 0 0 0 0 0 0 0 0 ...\n",
      " $ 2020.09.15.AML0024.AAACGCTAGCGAATGC           : num  0 0 0 0 0 0 0 0 0 0 ...\n",
      " $ 2020.09.15.AML0024.AAACGCTAGTTCCTGA           : num  0 0 0 0 0 ...\n",
      " $ 2020.09.15.AML0024.AAACGCTGTCTACAAC           : num  0 0 0 0 0 0 0 0 0 0 ...\n",
      " $ 2020.09.15.AML0024.AAAGAACAGCACCCAC           : num  0 0 0 0 0 0 0 0 0 0 ...\n",
      " $ 2020.09.15.AML0024.AAAGAACAGCGGACAT           : num  0 0 0 0 0 0 0 0 0 0 ...\n",
      " $ 2020.09.15.AML0024.AAAGAACAGCTGACTT           : num  0 0 0 0.647 0 ...\n",
      " $ 2020.09.15.AML0024.AAAGAACAGGGCTTCC           : num  0 0 0 0 0 ...\n",
      " $ 2020.09.15.AML0024.AAAGAACAGTAAGCAT           : num  0 0 0 0 0 0 0 0 0 0 ...\n",
      " $ 2020.09.15.AML0024.AAAGAACCAACAAAGT           : num  0 0 0 0 0 0 0 0 0 0 ...\n",
      " $ 2020.09.15.AML0024.AAAGAACGTACTCGCG           : num  0 0 0 0 0 ...\n",
      " $ 2020.09.15.AML0024.AAAGAACGTCATCGCG           : num  0 0.798 0 0 0 ...\n",
      " $ 2020.09.15.AML0024.AAAGAACGTGCATACT           : num  0 0 0 0 0 ...\n",
      " $ 2020.09.15.AML0024.AAAGAACGTTTCCAAG           : num  0 0 0 0 0 ...\n",
      " $ 2020.09.15.AML0024.AAAGAACTCCAAATGC           : num  0 0 0 0 0 0 0 0 0 0 ...\n",
      " $ 2020.09.15.AML0024.AAAGAACTCTGGGTCG           : num  0 0 0 0 0 0 0 0 0 0 ...\n",
      " $ 2020.09.15.AML0024.AAAGGATAGCCAGACA           : num  0 0 0 0 0 ...\n",
      " $ 2020.09.15.AML0024.AAAGGATAGCTAGCCC           : num  0 0 0 0 0 ...\n",
      " $ 2020.09.15.AML0024.AAAGGATAGGCTGGAT           : num  0 0 0 0 0 ...\n",
      " $ 2020.09.15.AML0024.AAAGGATAGTTGGCTT           : num  0 0.535 0 0 0 ...\n",
      " $ 2020.09.15.AML0024.AAAGGATCAGACCCGT           : num  0 0 0 0 0 0 0 0 0 0 ...\n",
      " $ 2020.09.15.AML0024.AAAGGATCATCGCTGG           : num  0 0 0 0 0 0 0 0 0 0 ...\n",
      " $ 2020.09.15.AML0024.AAAGGATCATTGGGAG           : num  0 0 0 0 0 0 0 0 0 0 ...\n",
      " $ 2020.09.15.AML0024.AAAGGATTCCCATACC           : num  0 0 0 0 0 ...\n",
      " $ 2020.09.15.AML0024.AAAGGGCCAATTAGGA           : num  0 0 0 0 0 ...\n",
      " $ 2020.09.15.AML0024.AAAGGGCGTGGTCTCG           : num  0 0 0 0 0 0 0 0 0 0 ...\n",
      " $ 2020.09.15.AML0024.AAAGGGCGTTGGGCCT           : num  0 0 0 0 0 0 0 0 0 0 ...\n",
      " $ 2020.09.15.AML0024.AAAGGGCTCATGCCCT           : num  0 0 0 0 0 0 0 0 0 0 ...\n",
      " $ 2020.09.15.AML0024.AAAGGTAAGGACTGGT           : num  0 0 0 0 0 0 0 0 0 0 ...\n",
      " $ 2020.09.15.AML0024.AAAGGTAAGGGAGGTG           : num  0 0 0 0 0 0 0 0 0 0 ...\n",
      " $ 2020.09.15.AML0024.AAAGGTACATCTCAAG           : num  0 0 0 0 0 0 0 0 0 0 ...\n",
      " $ 2020.09.15.AML0024.AAAGGTATCCTAAGTG           : num  0 0 0 0 0 0 0 0 0 0 ...\n",
      " $ 2020.09.15.AML0024.AAAGGTATCTCAGAAC           : num  0 0 0 0 0 0 0 0 0 0 ...\n",
      " $ 2020.09.15.AML0024.AAAGGTATCTGATTCT           : num  0 0 0 0 0 ...\n",
      " $ 2020.09.15.AML0024.AAAGTCCCACCAGTTA           : num  0 0 0 0 0 0 0 0 0 0 ...\n",
      " $ 2020.09.15.AML0024.AAAGTCCCAGAACGCA           : num  0 0 0 0 0 0 0 0 0 0 ...\n",
      " $ 2020.09.15.AML0024.AAAGTGAAGATCCGAG           : num  0 0 0 0 0 0 0 0 0 0 ...\n",
      " $ 2020.09.15.AML0024.AAAGTGAAGCGTGTCC           : num  0 0 0 0 0 0 0 0 0 0 ...\n",
      " $ 2020.09.15.AML0024.AAAGTGAAGGTGGTTG           : num  0 0 0 0 0 0 0 0 0 0 ...\n",
      " $ 2020.09.15.AML0024.AAAGTGAAGTCAGGGT           : num  0 0 0 0 0 0 0 0 0 0 ...\n",
      " $ 2020.09.15.AML0024.AAAGTGAAGTGAATAC           : num  0 0 0.598 0 0 ...\n",
      " $ 2020.09.15.AML0024.AAAGTGACAACCGCCA           : num  0 0 0 0 0 0 0 0 0 0 ...\n",
      " $ 2020.09.15.AML0024.AAAGTGAGTCAACCTA           : num  0 0 0 0 0 ...\n",
      " $ 2020.09.15.AML0024.AAAGTGAGTCGTTGCG           : num  0 0 0 0 0 0 0 0 0 0 ...\n",
      " $ 2020.09.15.AML0024.AAAGTGAGTTACACAC           : num  0 0 0 0 0 0 0 0 0 0 ...\n",
      " $ 2020.09.15.AML0024.AAAGTGATCATTCGTT           : num  0 0 0 0 0 0 0 0 0 0 ...\n",
      " $ 2020.09.15.AML0024.AAATGGAAGCTGTGCC           : num  0 0 0 0 0 0 0 0 0 0 ...\n",
      "  [list output truncated]\n",
      "NULL\n"
     ]
    }
   ],
   "source": [
    "# # load counts\n",
    "print(\"load counts\")\n",
    "counts <- fread(paste0(input_dir,\"counts_corr.csv.gz\"), header = TRUE)\n",
    "counts <- as.data.frame(counts)\n",
    "rownames(counts) <- counts$gene_symbol\n",
    "counts <- counts[,-1]\n",
    "print(str(counts))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "0c85040b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"load cell annotation\"\n",
      "'data.frame':\t46702 obs. of  93 variables:\n",
      " $ sample_ID                    : chr  \"AML-0024\" \"AML-0024\" \"AML-0024\" \"AML-0024\" ...\n",
      " $ cell                         : chr  \"2020-09-15-AML0024:CATCAAGGTTAGCGGA\" \"2020-09-15-AML0024:CATCAAGTCCGAGAAG\" \"2020-09-15-AML0024:CATCCACAGGGACCAT\" \"2020-09-15-AML0024:CCTCAACAGAGCAAGA\" ...\n",
      " $ UMAP_1                       : num  -0.731 -2.2 -2.867 -1.666 -0.972 ...\n",
      " $ UMAP_2                       : num  -15.8 -16.7 -16.1 -16.1 -17.5 ...\n",
      " $ orig.ident                   : chr  \"2020-09-15-AML0024\" \"2020-09-15-AML0024\" \"2020-09-15-AML0024\" \"2020-09-15-AML0024\" ...\n",
      " $ samples                      : chr  \"AML0024\" \"AML0024\" \"AML0024\" \"AML0024\" ...\n",
      " $ Broad_cell_identity          : chr  \"CD14+ monocyte\" \"CD14+ monocyte\" \"CD16+ monocyte\" \"CD14+ monocyte\" ...\n",
      " $ Cell_type_identity           : chr  \"CD14+ IFN+\" \"CD14+\" \"CD16+\" \"CD14+ IFN+\" ...\n",
      " $ clusters_res.2               : int  7 7 7 7 80 7 7 7 7 7 ...\n",
      " $ CNV_pos                      : chr  \"CNV+\" \"CNV+\" \"CNV+\" \"CNV+\" ...\n",
      " $ malignant                    : chr  \"malignant\" \"malignant\" \"malignant\" \"malignant\" ...\n",
      " $ aml                          : chr  \"AML\" \"AML\" \"AML\" \"AML\" ...\n",
      " $ ap_aml_age                   : chr  \"adult_AML\" \"adult_AML\" \"adult_AML\" \"adult_AML\" ...\n",
      " $ age_group                    : chr  \"adult_60up\" \"adult_60up\" \"adult_60up\" \"adult_60up\" ...\n",
      " $ sex.x                        : chr  \"male\" \"male\" \"male\" \"male\" ...\n",
      " $ inflammation_group           : chr  \"Inflammation_high\" \"Inflammation_high\" \"Inflammation_high\" \"Inflammation_high\" ...\n",
      " $ occupancy_score              : num  1 1 1 1 1 1 1 1 1 1 ...\n",
      " $ GSM                          : chr  \"GSM5613779\" \"GSM5613779\" \"GSM5613779\" \"GSM5613779\" ...\n",
      " $ date                         : chr  \"2020-09-15\" \"2020-09-15\" \"2020-09-15\" \"2020-09-15\" ...\n",
      " $ cell_ID.1                    : chr  \"2020.09.15.AML0024.CATCAAGGTTAGCGGA\" \"2020.09.15.AML0024.CATCAAGTCCGAGAAG\" \"2020.09.15.AML0024.CATCCACAGGGACCAT\" \"2020.09.15.AML0024.CCTCAACAGAGCAAGA\" ...\n",
      " $ cell_type_original           : chr  \"CD14+ monocyte\" \"CD14+ monocyte\" \"CD16+ monocyte\" \"CD14+ monocyte\" ...\n",
      " $ Sample.ID                    : chr  \"U-06-0024\" \"U-06-0024\" \"U-06-0024\" \"U-06-0024\" ...\n",
      " $ sex.y                        : chr  \"M\" \"M\" \"M\" \"M\" ...\n",
      " $ Age                          : int  64 64 64 64 64 64 64 64 64 64 ...\n",
      " $ Sequencing                   : chr  \"CITE-seq\" \"CITE-seq\" \"CITE-seq\" \"CITE-seq\" ...\n",
      " $ Institution                  : chr  \"OSU\" \"OSU\" \"OSU\" \"OSU\" ...\n",
      " $ Manuscript.ID                : chr  \"AML0024\" \"AML0024\" \"AML0024\" \"AML0024\" ...\n",
      " $ Dx                           : chr  \"AML\" \"AML\" \"AML\" \"AML\" ...\n",
      " $ DxELN_Cytogenetic            : chr  \"Adverse\" \"Adverse\" \"Adverse\" \"Adverse\" ...\n",
      " $ Cytogenetics                 : chr  \"48,XY,+8,+8,i(8)(p10),t(9;15)(q33;q15),del(10)(q22.1q24),del(13)(q14q21)\" \"48,XY,+8,+8,i(8)(p10),t(9;15)(q33;q15),del(10)(q22.1q24),del(13)(q14q21)\" \"48,XY,+8,+8,i(8)(p10),t(9;15)(q33;q15),del(10)(q22.1q24),del(13)(q14q21)\" \"48,XY,+8,+8,i(8)(p10),t(9;15)(q33;q15),del(10)(q22.1q24),del(13)(q14q21)\" ...\n",
      " $ Overall.survival             : num  137 137 137 137 137 137 137 137 137 137 ...\n",
      " $ Flow.report.summary          : chr  \"\" \"\" \"\" \"\" ...\n",
      " $ RAS.pathway                  : chr  \"\" \"\" \"\" \"\" ...\n",
      " $ NPM1                         : chr  \"\" \"\" \"\" \"\" ...\n",
      " $ IDH                          : chr  \"\" \"\" \"\" \"\" ...\n",
      " $ TET2                         : chr  \"\" \"\" \"\" \"\" ...\n",
      " $ TP53                         : chr  \"\" \"\" \"\" \"\" ...\n",
      " $ RUNX1                        : chr  \"\" \"\" \"\" \"\" ...\n",
      " $ CBF                          : logi  NA NA NA NA NA NA ...\n",
      " $ MLL.rearranged               : logi  NA NA NA NA NA NA ...\n",
      " $ IDH1                         : chr  \"0+T2:BL24\" \"0+T2:BL24\" \"0+T2:BL24\" \"0+T2:BL24\" ...\n",
      " $ IDH2                         : chr  \"0,46\" \"0,46\" \"0,46\" \"0,46\" ...\n",
      " $ ASXL1                        : chr  \"0\" \"0\" \"0\" \"0\" ...\n",
      " $ BCOR                         : chr  \"0,99\" \"0,99\" \"0,99\" \"0,99\" ...\n",
      " $ CBL                          : num  0 0 0 0 0 0 0 0 0 0 ...\n",
      " $ CEBPAdm                      : num  0 0 0 0 0 0 0 0 0 0 ...\n",
      " $ CSF3R                        : chr  \"0\" \"0\" \"0\" \"0\" ...\n",
      " $ DNMT3A                       : chr  \"0\" \"0\" \"0\" \"0\" ...\n",
      " $ ETV6                         : num  0 0 0 0 0 0 0 0 0 0 ...\n",
      " $ FLT3.TKD                     : chr  \"0,06\" \"0,06\" \"0,06\" \"0,06\" ...\n",
      " $ FLT3.ITD                     : num  0 0 0 0 0 0 0 0 0 0 ...\n",
      " $ GATA2                        : num  0 0 0 0 0 0 0 0 0 0 ...\n",
      " $ JAK2                         : num  0 0 0 0 0 0 0 0 0 0 ...\n",
      " $ KIT                          : num  0 0 0 0 0 0 0 0 0 0 ...\n",
      " $ NPM1.1                       : chr  \"0,17\" \"0,17\" \"0,17\" \"0,17\" ...\n",
      " $ NRAS                         : num  0 0 0 0 0 0 0 0 0 0 ...\n",
      " $ KRAS                         : chr  \"0\" \"0\" \"0\" \"0\" ...\n",
      " $ PTPN11                       : chr  \"0,17\" \"0,17\" \"0,17\" \"0,17\" ...\n",
      " $ PHF6                         : num  0 0 0 0 0 0 0 0 0 0 ...\n",
      " $ RAD21                        : num  0 0 0 0 0 0 0 0 0 0 ...\n",
      " $ RUNX1.1                      : chr  \"0\" \"0\" \"0\" \"0\" ...\n",
      " $ SETBP1                       : num  0 0 0 0 0 0 0 0 0 0 ...\n",
      " $ SMC1A                        : num  0 0 0 0 0 0 0 0 0 0 ...\n",
      " $ STAG2                        : num  0 0 0 0 0 0 0 0 0 0 ...\n",
      " $ SRSF2                        : chr  \"0\" \"0\" \"0\" \"0\" ...\n",
      " $ SF3B1                        : num  0 0 0 0 0 0 0 0 0 0 ...\n",
      " $ TET2.1                       : chr  \"0\" \"0\" \"0\" \"0\" ...\n",
      " $ U2AF1                        : chr  \"0,45\" \"0,45\" \"0,45\" \"0,45\" ...\n",
      " $ WT1                          : num  0 0 0 0 0 0 0 0 0 0 ...\n",
      " $ TP53.1                       : chr  \"0\" \"0\" \"0\" \"0\" ...\n",
      " $ ZRSR2                        : num  0 0 0 0 0 0 0 0 0 0 ...\n",
      " $ health_status                : chr  \"AML\" \"AML\" \"AML\" \"AML\" ...\n",
      " $ case_or_control              : chr  \"case\" \"case\" \"case\" \"case\" ...\n",
      " $ cell_type                    : chr  \"Mono\" \"Mono\" \"Mono\" \"Mono\" ...\n",
      " $ cell_subtype                 : chr  \"Mono\" \"Mono\" \"Mono\" \"Mono\" ...\n",
      " $ total_counts_raw             : num  2591 2598 2665 2751 2845 ...\n",
      " $ mito_counts                  : num  31.5 37.5 40.8 37.1 44.3 ...\n",
      " $ frac_mito                    : num  0.0122 0.0144 0.0153 0.0135 0.0156 ...\n",
      " $ nr_genes                     : int  1946 2177 2027 2731 2161 2297 2783 2352 2691 2659 ...\n",
      " $ counts_upper_threshold       : int  30000 30000 30000 30000 30000 30000 30000 30000 30000 30000 ...\n",
      " $ counts_lower_threshold       : int  1100 1100 1100 1100 1100 1100 1100 1100 1100 1100 ...\n",
      " $ genes_lower_threshold        : int  500 500 500 500 500 500 500 500 500 500 ...\n",
      " $ counts_upper_filter          : chr  \"True\" \"True\" \"True\" \"True\" ...\n",
      " $ counts_lower_filter          : chr  \"True\" \"True\" \"True\" \"True\" ...\n",
      " $ genes_lower_filter           : chr  \"True\" \"True\" \"True\" \"True\" ...\n",
      " $ cell_filter                  : chr  \"True\" \"True\" \"True\" \"True\" ...\n",
      " $ threshold_celltype_size      : int  5 5 5 5 5 5 5 5 5 5 ...\n",
      " $ celltype_size_cf             : int  5451 5451 5451 5451 5451 5451 5451 5451 5451 452 ...\n",
      " $ celltype_size_filter         : chr  \"True\" \"True\" \"True\" \"True\" ...\n",
      " $ cell_type_ID                 : chr  \"Mono_AML-0024\" \"Mono_AML-0024\" \"Mono_AML-0024\" \"Mono_AML-0024\" ...\n",
      " $ total_counts_raw_cf_ctf_gf   : num  2584 2594 2656 2743 2843 ...\n",
      " $ total_counts_raw_cf_ctf_gf_sf: num  2584 2594 2656 2743 2843 ...\n",
      " $ bares_mutation               : chr  \"malignant\" \"malignant\" \"malignant\" \"malignant\" ...\n",
      "NULL\n"
     ]
    }
   ],
   "source": [
    "# load cell annotation\n",
    "print(\"load cell annotation\")\n",
    "anno_cells <- read.table(paste0(input_dir,\"anno_cells_corr.txt\")\n",
    "                         ,sep = \"\\t\"\n",
    "                         ,row.names = 1\n",
    "                         ,header = TRUE\n",
    "                         )\n",
    "print(str(anno_cells))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "416648ee",
   "metadata": {},
   "outputs": [],
   "source": [
    "colnames(counts) <- anno_cells$cell_ID\n",
    "rownames(anno_cells) <- anno_cells$cell_ID"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "ab171401",
   "metadata": {},
   "outputs": [],
   "source": [
    "seurat_obj=CreateSeuratObject(counts=counts, meta.data=anno_cells)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "c34ba1f5",
   "metadata": {},
   "outputs": [],
   "source": [
    "Idents(seurat_obj) <- \"cell_type\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "8490cace",
   "metadata": {},
   "outputs": [],
   "source": [
    "cell_type_list <- unique(seurat_obj@meta.data$cell_type)\n",
    "sample_list <- unique(seurat_obj@meta.data$sample_ID)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1bdade5b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "2a5a02c6",
   "metadata": {},
   "outputs": [],
   "source": [
    "cell_types <- unique(anno_cells[,\"cell_type\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "f9ab4b59",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_mean_above_threshold <- matrix(,nrow = nrow(counts)\n",
    "                                          ,ncol = length(cell_types)\n",
    "        )\n",
    "rownames(df_mean_above_threshold) <- rownames(counts)\n",
    "colnames(df_mean_above_threshold) <- cell_types"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "129e12ca",
   "metadata": {},
   "outputs": [],
   "source": [
    "threshold_expr=0.1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "f075c879",
   "metadata": {},
   "outputs": [],
   "source": [
    "for(cell_type in cell_types){\n",
    "                # identify cells that belong to the cell_type\n",
    "                idx_cell_type <- anno_cells[,\"cell_type\"] == cell_type\n",
    "                \n",
    "                # subset counts to the cell type of interest\n",
    "                counts_sub <- counts[,idx_cell_type]\n",
    "                \n",
    "                # identify counts that pass the threshold\n",
    "                above_threshold <- counts_sub > threshold_expr\n",
    "                \n",
    "                # set all counts which did not pass the threshold to zero\n",
    "                counts_sub_thr <- counts_sub * above_threshold\n",
    "                \n",
    "                # populate the df_mean_above_threshold\n",
    "                ifelse(is.null(dim(counts_sub_thr))\n",
    "                       ,df_mean_above_threshold[,cell_type] <- counts_sub_thr\n",
    "                       ,{\n",
    "                               # substitute zeros with NAs\n",
    "                               counts_sub_thr[counts_sub_thr == 0] <- NA\n",
    "                               df_mean_above_threshold[,cell_type] <- rowMeans(counts_sub_thr\n",
    "                                                                               ,na.rm = TRUE)\n",
    "                       }\n",
    "                       \n",
    "                )\n",
    "                # subsitute NAs with zeroes\n",
    "                df_mean_above_threshold[is.na(df_mean_above_threshold)] <- 0\n",
    "                \n",
    "        }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7705c654",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "76a03253",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0ca54099",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0ab1f8da",
   "metadata": {},
   "outputs": [],
   "source": [
    "# define all cell typess in the data\n",
    "        cell_types <- unique(anno_cells[,\"cell_type\"])\n",
    "        \n",
    "        # create empty dataframe nr_ligands_receptors x nr_cell_types\n",
    "        # we will populate it the the values of the mean expression values\n",
    "        df_mean_above_threshold <- matrix(,nrow = nrow(counts)\n",
    "                                          ,ncol = length(cell_types)\n",
    "        )\n",
    "        rownames(df_mean_above_threshold) <- rownames(counts)\n",
    "        colnames(df_mean_above_threshold) <- cell_types\n",
    "        \n",
    "        # populate the df_mean_above_threshold\n",
    "        for(cell_type in cell_types){\n",
    "                # identify cells that belong to the cell_type\n",
    "                idx_cell_type <- anno_cells[,\"cell_type\"] == cell_type\n",
    "                \n",
    "                # subset counts to the cell type of interest\n",
    "                counts_sub <- counts[,idx_cell_type]\n",
    "                \n",
    "                # identify counts that pass the threshold\n",
    "                above_threshold <- counts_sub > threshold_expr\n",
    "                \n",
    "                # set all counts which did not pass the threshold to zero\n",
    "                counts_sub_thr <- counts_sub * above_threshold\n",
    "                \n",
    "                # populate the df_mean_above_threshold\n",
    "                ifelse(is.null(dim(counts_sub_thr))\n",
    "                       ,df_mean_above_threshold[,cell_type] <- counts_sub_thr\n",
    "                       ,{\n",
    "                               # substitute zeros with NAs\n",
    "                               counts_sub_thr[counts_sub_thr == 0] <- NA\n",
    "                               df_mean_above_threshold[,cell_type] <- rowMeans(counts_sub_thr\n",
    "                                                                               ,na.rm = TRUE)\n",
    "                       }\n",
    "                       \n",
    "                )\n",
    "                # subsitute NAs with zeroes\n",
    "                df_mean_above_threshold[is.na(df_mean_above_threshold)] <- 0\n",
    "                \n",
    "        }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6adda7e7",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.1.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
